lib/Dialect/StaticLogic/StaticLogicOps.cpp

```diff
@@ -180,6 +183,59 @@ void PipelineWhileOp::build(OpBuilder &builder, OperationState &state,
                                        ValueRange());
 }
 
+/// Get the terminator of the pipeline.
+PipelineTerminatorOp PipelineWhileOp::getTerminator() {
+  return cast<PipelineTerminatorOp>(getStagesBlock().getTerminator());
+}
+
+/// Compute the distance between the two stages.
+uint64_t PipelineWhileOp::computeDistance(PipelineStageOp from,
+                                          PipelineStageOp to) {
+  // Ensure these are valid stages within this pipeline.
+  assert(from && to && from->getParentOp() == getOperation() &&
+         to->getParentOp() == getOperation());
+
+  // Traverse the linked list to determine the distance.
+  uint64_t distance = 0;
+  Operation *current = from;
+  while (current != to) {
+    current = from->getNextNode();
+    ++distance;
+  }
+
+  return distance;
+}
+
+/// Compute the initiation interval based on the iter args defs and uses.
+uint64_t PipelineWhileOp::computeInitiationInterval() {
+  uint64_t maxDistance = 0;
+
+  // Check the uses of the pipeline's block iter args, and the definitions of
+  // the pipeline's terminator iter args.
+  auto useIterArgs = getStagesBlock().getArguments();
+  auto defIterArgs = getTerminator().iter_args();
+  assert(useIterArgs.size() == defIterArgs.size());
+
+  for (size_t i = 0; i < useIterArgs.size(); ++i) {
+    Value useIterArg = useIterArgs[i];
+    Value defIterArg = defIterArgs[i];
+
+    // For each iter arg, find the maximum distance from the stage in which it
+    // is defined to latest the stage in which it is used.
+    auto definingStage = defIterArg.getDefiningOp<PipelineStageOp>();
+    for (auto &use : useIterArg.getUses()) {
+      auto usingStage = use.getOwner()->getParentOfType<PipelineStageOp>();
+      uint64_t distance = computeDistance(definingStage, usingStage);
```

Comment:
This is only correct for single-cycle operations! In general, the defining stage needs to be the one in which an op finishes.

Reply:
I think your example has an RecMII of 1, because the use of the *previous* iteration's multiplication result (via `%arg0`) is after the *current* iteration's product is computed, so that recurrence is always trivially satisfied.

> this is how I imagine we would represent a 3-cycle multiply. Is that correct?

Yup, that's what I remember from our last discussion: Returning values from a stage tells the lowering that is has to figure out the right amount of registers to preserve values until they're used. This would work because the lowering is aware of the latency of the operators.

The problem is that the pipeline-op-verifier doesn't know in which stage the values become available. I think we would want `%0` to behave like a [future](https://en.wikipedia.org/wiki/Futures_and_promises), but as far as MLIR is concerned, it's fine e.g. to use `%0` in stage 1.

Reply:
That all makes sense.

I'm try to back up to how we defined the II rules initially. The [doc](https://circt.llvm.org/docs/Dialects/StaticLogic/#staticlogicpipelinewhile-circtstaticlogicpipelinewhileop) I wrote says:

> The maximum distance in stages between where an iter_arg is used and where that iter_arg is registered must be less than the loop’s initiation interval. For example, with II=1, each iter_arg must be used and registered in the same stage.

I don't think I was really considering multiple cycle operations when I wrote that. The logic we are discussing here is me trying to implement that idea directly.

Maybe there is a simpler rule, given the current, simple pipeline op (no latency or start time attributes). Could we just use the index (in the list of stages) of the latest stage to yield an iter_arg is the II? In the simple loops we've considered with II=1, we'd yield the iter_arg from the first stage. If an iter arg was yielded from the second stage, we would have II=2.

I'm trying to dig up my II>1 examples to remember how we'd schedule them currently.

Reply:
For example, I have this program:

```mlir
func @II2(%arg0: memref<?xi32>) {
  affine.for %arg1 = 2 to 10 {
    %0 = affine.load %arg0[%arg1 - 2] : memref<?xi32>
    %1 = affine.load %arg0[%arg1 - 1] : memref<?xi32>
    %2 = arith.addi %0, %1 : i32
    affine.store %2, %arg0[%arg1] : memref<?xi32>
  }
  return
}
```

The scheduling result is:

```
Scheduled initiation interval = 2

Scheduling outputs for %0 = affine.load %arg0[%arg1 - 2] : memref<?xi32>
  start = 0

Scheduling outputs for %1 = affine.load %arg0[%arg1 - 1] : memref<?xi32>
  start = 0

Scheduling outputs for %2 = arith.addi %0, %1 : i32
  start = 1

Scheduling outputs for affine.store %2, %arg0[%arg1] : memref<?xi32>
  start = 1
```

Reply:
> part of me is starting to think we should just require this attribute be specified from the pass/tool/etc. that creates this operation

We just discussed this, and the bottom line is we really expect a scheduling problem to be solved before this operation is created. I will try to capture what we discussed in the operation's documentation, and we can review that.